pkg/controller/worker/reconciler.go

```diff
@@ -84,91 +84,166 @@ func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
+		worker.Status.LastOperation = extensionscontroller.LastOperation(lastOperationType, gardencorev1beta1.LastOperationStateProcessing, 1, description)
+		return nil
+	})
+}
 
-		msg := "Successfully deleted worker"
-		r.logger.Info(msg, "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
-		if err := r.updateStatusSuccess(r.ctx, worker, operationType, msg); err != nil {
-			return reconcile.Result{}, err
-		}
+func (r *reconciler) updateStatusError(ctx context.Context, err error, worker *extensionsv1alpha1.Worker, lastOperationType gardencorev1beta1.LastOperationType, description string) {
+	r.logger.Error(err, description, "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+	updateErr := extensionscontroller.TryUpdateStatus(ctx, retry.DefaultBackoff, r.client, worker, func() error {
+		worker.Status.ObservedGeneration = worker.Generation
+		worker.Status.LastOperation, worker.Status.LastError = extensionscontroller.ReconcileError(lastOperationType, gardencorev1beta1helper.FormatLastErrDescription(fmt.Errorf("%s: %v", description, extensionscontroller.ReconcileErrCauseOrErr(err))), 50, gardencorev1beta1helper.ExtractErrorCodes(err)...)
+		return nil
+	})
+	utilruntime.HandleError(updateErr)
+}
 
-		r.logger.Info("Removing finalizer.", "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
-		if err := extensionscontroller.DeleteFinalizer(r.ctx, r.client, FinalizerName, worker); err != nil {
-			r.logger.Error(err, "Error removing finalizer from Worker", "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
-			return reconcile.Result{}, err
-		}
+func (r *reconciler) updateStatusSuccess(ctx context.Context, worker *extensionsv1alpha1.Worker, lastOperationType gardencorev1beta1.LastOperationType, description string) error {
+	r.logger.Info(description, "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+	return extensionscontroller.TryUpdateStatus(ctx, retry.DefaultBackoff, r.client, worker, func() error {
+		worker.Status.ObservedGeneration = worker.Generation
+		worker.Status.LastOperation, worker.Status.LastError = extensionscontroller.ReconcileSucceeded(lastOperationType, description)
+		return nil
+	})
+}
+
+func (r *reconciler) removeFinalizerFromWorker(worker *extensionsv1alpha1.Worker) error {
+	r.logger.Info("Removing finalizer.", "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+	if err := extensionscontroller.DeleteFinalizer(r.ctx, r.client, FinalizerName, worker); err != nil {
+		r.logger.Error(err, "Error removing finalizer from Worker", "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+		return err
+	}
+	return nil
+}
+
+func (r *reconciler) migrate(worker *extensionsv1alpha1.Worker, cluster *extensionscontroller.Cluster) (reconcile.Result, error) {
+	// remove operation annotation 'migrate'
+	withOpAnnotationRestore := worker.DeepCopyObject()
+	delete(worker.Annotations, v1beta1constants.GardenerOperation)
```

Comment:
do I understand that correctly that the Migration annotation is removed before the migration operation has completed? Why is it not done after the successful migration?

Reply:
We had some talks regarding that (wrt to the `restore` annotation) and we will try to make the `migrate` annotation behave similarly - so that it is removed at the end.

Reply:
Ok, I will put it at the end